course3/homework-uniqueptr/test.cc

```diff
@@ -0,0 +1,20 @@
+#include "unique_ptr.h"
+#include <iostream>
+#include <memory>
+#include <cstring>
+
+//template<typename template<typename T> class PtrType>
+template <typename PtrType>
+void test(){
+	PtrType a(new char[5]);
+	std::strcpy(a.get(),"abcd");
+	std::cout << "value of a=" << a.get() <<std::endl;	
+}
+
+int main(){
+	std::cout<<"test std::unique_ptr"<<std::endl;
+	test<std::unique_ptr<char[]>>();
+	std::cout<<std::endl;
+	std::cout<<"test MyUniquePtr"<<std::endl;
+	test<MyUniquePtr<char[]>>();
```

Comment:
Hi Maciek, looking for some advice

at compile time I'm getting the following which works for std::unique_ptr
test.cc:9:23: error: no matching function for call to ‘MyUniquePtr<char []>::MyUniquePtr(char*)
The type I instantiate MyUniquePtr with is char [],  the call to (new char[5]) returns type char * so I see why this is happening, but I can't think of any solution, I would hope that the compiler could kind of ignore the difference between char* and char[]. My suspicion is maybe there is no neat solution for arrays and I would need to add specialisation for that case, I think I remember you saying something about that.

On another note, I was trying to structure my test function to just take the class template MyUniquePtr, and delved into the concept of class template templates. Couldn't really get it working so I just decided to pass the instantiated UniquePtr<char[]> to the test function template instead. Would you be able to off any advice on how to set up tests for template classes?